### PR TITLE
[Feature] Added heterograph support to SDDMM_COO and clean up SpMM and SDDMM hetero kernels

### PR DESCRIPTION
**Pull request by isratnisa**
## Description
Added heterogeneous graph support to SDDMM kernel using COO storage format. SDDMM using COO format uses edge level parallelism and is much faster than using CSR format based SDDMM. The default format is now set to COO format which is preferred format for homogeneous cases too.

Also, cleaned up some code in  SpMM and SDDMM kernels, e.g.,`cuda::SpMMCsrHetero` and `cuda::SDDMMCsrHetero` kernels which was initially written to support CUDA multi stream.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by nv-dlasalle — APPROVED** [review]
Looks good to me. Is SDDMMHetero covered by the regression test? If so, can we run that before merging?

**Comment by BarclayII** [comment]
@dgl-bot run c5.9xlarge api dmlc/master

File: benchmarks/benchmarks/api/bench_builtin_apply_edges_hetero.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+@utils.parametrize('num_relations', [5, 50, 500])
+@utils.parametrize('format', ['coo', 'csr'])
+@utils.parametrize('feat_size', [8, 128, 512])
+@utils.parametrize('reduce_type', ['u->e']) #, 'e->u'])
+
+def track_time( num_relations, format, feat_size, reduce_type):
+    device = utils.get_bench_device()
+        dd = {}
+    candidate_edges = [dgl.data.CoraGraphDataset(verbose=False)[0].edges(), dgl.data.PubmedGraphDataset(verbose=False)[
+        0].edges(), dgl.data.CiteseerGraphDataset(verbose=False)[0].edges()]
```
Comment (zheng-da) [inline benchmarks/benchmarks/api/bench_builtin_apply_edges_hetero.py:20]: we should develop a proper workload for testing the heterogeneous graph API.
the performance improvement or overhead cannot really be seen in this dataset. we should include the benchmark of the knowledge graphs.

**Comment by BarclayII** [comment]
@dgl-bot run c5.9xlarge api.bench_builtin_apply_edges_hetero dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run c5.9xlarge api.bench_builtin_apply_edges_hetero dmlc/master

File: benchmarks/benchmarks/api/bench_builtin_apply_edges_hetero.py (hunk 1 above)
Comment (BarclayII) [reply benchmarks/benchmarks/api/bench_builtin_apply_edges_hetero.py:20]: Will do that in later PRs.

**Review by BarclayII — APPROVED** [review]

**Pull request merged**